editor/src/core/workers/parser-printer/uid-fix.ts

```diff
@@ -0,0 +1,196 @@
+import { MapLike } from 'typescript'
+import {
+  isJSXElement,
+  isUtopiaJSXComponent,
+  JSXElementChild,
+  TopLevelElement,
+  UtopiaJSXComponent,
+} from '../../shared/element-template'
+import { optionalMap } from '../../shared/optional-utils'
+import { isParseSuccess, ParsedTextFile, StaticElementPath } from '../../shared/project-file-types'
+import * as TP from '../../shared/template-path'
+import { setUtopiaIDOnJSXElement } from '../../shared/uid-utils'
+import { getUtopiaID, transformJSXComponentAtElementPath } from '../../model/element-template-utils'
+import { getComponentsFromTopLevelElements } from '../../model/project-file-utils'
+import { mapArrayToDictionary } from '../../shared/array-utils'
+import { emptySet } from '../../shared/set-utils'
+import { fastForEach } from '../../shared/utils'
+
+// TODO move this function from editor-sate to a shared utils file, delete this copy-paste
+export function applyUtopiaJSXComponentsChanges(
+  topLevelElements: Array<TopLevelElement>,
+  newUtopiaComponents: Array<UtopiaJSXComponent>,
+): Array<TopLevelElement> {
+  // Run through the old top level elements, replacing the exported elements with those in the
+  // newly updated result with the same name.
+  // If it doesn't exist in the updated result, delete it.
+  // For any new items in the updated result, add them in.
+  const addedSoFar: Set<string> = emptySet()
+  let newTopLevelElements: Array<TopLevelElement> = []
+  fastForEach(topLevelElements, (oldTopLevelElement) => {
+    if (isUtopiaJSXComponent(oldTopLevelElement)) {
+      const updatedElement = newUtopiaComponents.find((e) => e.name === oldTopLevelElement.name)
+      if (updatedElement !== undefined) {
+        addedSoFar.add(updatedElement.name)
+        newTopLevelElements.push(updatedElement)
+      }
+    } else {
+      newTopLevelElements.push(oldTopLevelElement)
+    }
+  })
+
+  fastForEach(newUtopiaComponents, (updatedElement) => {
+    if (!addedSoFar.has(updatedElement.name)) {
+      newTopLevelElements.push(updatedElement)
+    }
+  })
+
+  return newTopLevelElements
+}
+
+export function fixParseSuccessUIDs(
+  oldParsed: ParsedTextFile | null,
+  newParsed: ParsedTextFile,
+): ParsedTextFile {
+  if (oldParsed == null || !isParseSuccess(oldParsed) || !isParseSuccess(newParsed)) {
+    // we won't try to fix parse failures
+    return newParsed
+  }
+
+  let newToOldUidMapping: {
+    [newUID: string]: {
+      oldUID: string
+      newUID: string
+      pathToModify: StaticElementPath
+    }
+  } = {}
+
+  zipProjectContents(
+    oldParsed.topLevelElements,
+    newParsed.topLevelElements,
+    (oldElement, newElement, newPath) => {
+      const oldUID = optionalMap(getUtopiaID, oldElement)
+      const newUID = optionalMap(getUtopiaID, newElement)
+
+      if (oldUID != null && newUID != null && oldUID !== newUID) {
+        // we have a UID mismatch
+        newToOldUidMapping[newUID] = { oldUID: oldUID, newUID: newUID, pathToModify: newPath }
+      }
```

Comment:
I think at this point we really need to check that we're actually matching up the elements correctly, otherwise we're going to be blinding assigning incorrect UIDs during duplications etc. which definitely risks displaying or using incorrect cached data. If we had the before and after code in scope we could probably use the Hamming distance to determine that

Reply:
I opted for a much simpler heuristic, (in line 55) if we found more than one mismatched UIDs, we just bail out.

Reply:
Does that mean if an element is duplicated it will trip that? I think that intended behaviour makes sense.